DM-257 CRUD resource actions for CWU proposals

sutherlanda opened the pull request:
This PR includes resource CRUD actions for CWU proposals.

sutherlanda commented:
This is ready for review @dhruvio

dhruvio commented on src/back-end/lib/resources/proposal/code-with-us.ts:
```diff
@@ -0,0 +1,403 @@
+              attachments: getStringArray(value, 'attachments')
+            });
+          case 'submit':
+            return adt('submit', String(value));
```
Change `String(value)` to `getString(body, 'value', '')`. If value is `undefined` or `null`, the current approach will simply store it as `"undefined"` or `"null"`.

dhruvio replied:
Same for the cases below.

dhruvio commented on src/shared/lib/resources/proposal/code-with-us.ts:
```diff
@@ -2,9 +2,10 @@ import { CWUOpportunitySlim } from 'shared/lib/resources/code-with-us';
+import { ErrorTypeFrom } from 'shared/lib/validation';
 
-export enum ProposalStatus {
+export enum CWUProposalStatus {
```
This needs another status — `Evaluated` for when a score has been entered.

dhruvio commented on src/shared/lib/resources/proposal/code-with-us.ts:
```diff
@@ -2,9 +2,10 @@ import { CWUOpportunitySlim } from 'shared/lib/resources/code-with-us';
+export enum CWUProposalStatus {
   Draft = 'DRAFT',
   Submitted = 'SUBMITTED',
   Review = 'REVIEW',
```
Is it possible to rename this to `UnderReview`?

dhruvio replied:
Status changes to `UnderReview` will happen automatically when a proposal deadline for an opportunity lapses. We should talk about how to handle this. Either via a daemon, or on-demand.

dhruvio commented on src/back-end/lib/resources/proposal/code-with-us.ts:
```diff
@@ -0,0 +1,403 @@
+              body: adt('submit', validatedSubmissionNote.value)
+            });
+          case 'score':
+            if (!isValidStatusChange(validatedCWUProposal.value.status, CWUProposalStatus.Review)) {
```
The `to` status should be `Evaluated`.

dhruvio commented on src/back-end/lib/resources/proposal/code-with-us.ts:
```diff
@@ -0,0 +1,403 @@
+          return invalid({ notFound: ['Proposal not found.'] });
+        }
+        if (validatedCWUProposal.value.status !== CWUProposalStatus.Draft) {
+          return invalid({ permissions: [permissions.ERROR_MESSAGE] });
```
Change this error message:

`{ status: ['You can not delete a proposal that is not a draft.'] }`

dhruvio commented on src/back-end/lib/db.ts:
```diff
@@ -1080,3 +1102,457 @@ export const deleteCWUOpportunity = tryDb<[Id], CWUOpportunity>(async (connectio
+        proposal: proposalId,
+        createdAt: now,
+        createdBy: session.user.id,
+        status: CWUProposalStatus.Review,
```
Change to `Evaluated`.

sutherlanda commented on src/back-end/lib/validation.ts:
```diff
@@ -117,3 +120,44 @@ export async function validateCWUOpportunityId(connection: db.Connection, opport
+      if (isValid(validatedIndividualProponentRequestBody)) {
+        return valid(adt('individual', validatedIndividualProponentRequestBody.value));
+      }
+      return invalid(adt('individual', validatedIndividualProponentRequestBody.value));
```
Had to add a `as const` but it typechecks :)

sutherlanda commented:
@dhruvio I've implemented the requested changes.  I've also added `statusHistory` for proposal responses to this branch.  `statusHistory` for opportunities is in `feature/andrew/cwu-opportunity-statuses`, but let me know if you want everything here.
